src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -225,10 +341,25 @@ void LMNNFunction<MetricType>::Gradient(const arma::mat& transformation,
       for (size_t l = 0, bp = k; l < bp ; l++)
       {
         // Calculate gradient due to triplets.
-        double eval = metric.Evaluate(transformedDataset.col(i),
-                          transformedDataset.col(targetNeighbors(j, i))) -
-                      metric.Evaluate(transformedDataset.col(i),
-                          transformedDataset.col(impostors(l, i)));
+        double eval = 0;
+
+        // Flag to trigger exact eval calculation.
+        bool exactEval = true;
+
+        // Use eval calualated during Evaluate.
+        if (!std::isnan(evalOld(l, j, i)))
```

Comment:
I think I have at least pinpointed the cause of difference in result - 

1.  In this PR we have breaking condition as `if (eval <= -1)` whereas in the current LMNN we have if `(eval < -1)`. Doing this made solved the iris deviation.

2.  There is something going wrong here while using cache evalOld for Gradient() function. If I change if (!std::isnan(evalOld(l, j, i))) to false, BBSGD deviation problem solves as well.

Reply:
> 1. In this PR we have breaking condition as if (eval <= -1) whereas in the current LMNN we have if (eval < -1). Doing this made solved the iris deviation. So, I think we don't need to worry about this.

Got it, so you mean that when you run the current PR with `if (eval < -1)`, you get the exact same results?

> 2. There is something going wrong here while using cache evalOld for Gradient() function. If I change if (!std::isnan(evalOld(l, j, i))) to false, BBSGD deviation problem solves as well.

Sorry, do you mean you change to `if (false)` or `if (std::isnan(evalOld(l, j, i)) == false)` or something else?  I don't know if I understood.

Reply:
> Got it, so you mean that when you run the current PR with if (eval < -1), you get the exact same results with L-BFGS on the iris dataset as the current git master branch?

Yes

> Sorry, do you mean you change to if (false) or if (std::isnan(evalOld(l, j, i)) == false) or something else? I don't know if I understood.Sorry, do you mean you change to if (false) or if (std::isnan(evalOld(l, j, i)) == false) or something else? I don't know if I understood.

Ya, I wrote a confusing statement. Sorry, about that. I meant  if (false)

Reply:
Here's is my thought about eval <= -1 
Our original condition was max(0, 1 + eval), and 0 doesn't contribute anything so skipping it shouldn't be any problem. And at eval = -1, we are basically  dealing with 0. Does it sounds right?

Ya, I will try that.

Reply:
You're right, when `eval == -1`, there is no change to the cost function.  But there is a change to the gradient:

```
         // Caculate gradient due to impostors.
         arma::vec diff = dataset.col(i) - dataset.col(targetNeighbors(j, i));
         cil += diff * arma::trans(diff);
 
         diff = dataset.col(i) - dataset.col(impostors(l, i));
         cil -= diff * arma::trans(diff);
```

In the git master branch, those terms get added to `cil` even if `eval == -1`.  When I review the paper it's not fully clear whether we should include those terms in our gradient computation.  The paper writes:

"Let us define a set of triples N_t, such that (i, j,l) ∈ N_t if and only if the indices (i, j,l) trigger the hinge loss in the second part of Eq. (18)."

I think we can go either way here, and I like the tighter condition of `if (eval <= -1)` since that will allow a little bit of extra pruning.  So I would say, let's leave it as it is with `<=`, but just in order to test correctness with BigBatchSGD, let's make sure we do that test with `<` to match the git master code.